python/dlr/counter/counter_mgr.py

```diff
@@ -0,0 +1,168 @@
+import platform
+import atexit
+import json
+import os
+import logging
+import threading
+
+from .publisher import MsgPublisher
+from .system import Factory
+from . import config
+from .model_exec_counter import ModelExecCounter
+from .model_metric import ModelMetric
+from .utils.helper import *
+
+
+def call_home(func):
+    def wrapped_call_home(*args, **kwargs):
+        call_counter = CallCounterMgr.get_instance()
+        if call_counter is not None:
+            if func.__name__ == "init_call_home":
+                print(config.CALL_HOME_USR_NOTIFICATION)
+                func(*args, **kwargs)
+                call_counter.runtime_loaded()
+            elif func.__name__ == "__init__":
+                func(*args, **kwargs)
+                obj = args[0]
+                call_counter.model_loaded(obj.get_model_name())
+            else:
+                res = func(*args, **kwargs)
+                obj = args[0]
+                call_counter.model_run(obj.get_model_name())
+                return res
+
+    return wrapped_call_home
+
+
+getinst = threading.Lock()
+relinst = threading.Lock()
+
+
+class CallCounterMgr(object):
+    RUNTIME_LOAD = 1
+    MODEL_LOAD = 2
+    MODEL_RUN = 3
+    _instance = None
+    DEFAULT_ENABLE_FEATURE = True
+
+    @staticmethod
+    def get_instance():
+        """return single instance of class"""
+        with getinst:
+            if CallCounterMgr._instance is None:
+                if CallCounterMgr.is_feature_enabled():
+                    try:
+                        CallCounterMgr._instance = CallCounterMgr()
+                    except Exception as e:
+                        CallCounterMgr._instance = None
+                        logging.exception("unsupported system for call home feature")
+                    else:
+                        atexit.register(CallCounterMgr._instance.stop)
+                else:
+                    logging.warning("call home feature disabled")
+            return CallCounterMgr._instance
+
+    def __init__(self):
+        try:
+            machine_typ = platform.machine()
+            os_name = platform.system()
+            os_supt = "{0}_{1}".format(os_name, machine_typ)
+            self.system = Factory.get_system(os_supt)
+            if self.system is None:
+                raise Exception("unsupported system")
```

Comment:
This is a bug. We're shouldn't throw exception and stop DLR because we don't support phone-home. The reason is that your system validation will be expired someday in the future; however, that doesn't mean DLR won't work. Instead, we should log the warning and disable the feature. In addition, I suggest that we should have the validation check prior initialize CallCounterMgr.  Something similar to `is_feature_enabled`, we can have something like `is_feature_supported` and log the warning there.

Following is an exception to indicate this problem. I'm testing this on imx8 device with Linux_aarch64. DLR will work, but because of the version problem, DLR was forced to stop. 

[Errno 2] No such file or directory: '/opt/ml/errors/errors.log'
ERROR:root:while in counter mgr init
Traceback (most recent call last):
  File "/home/torizon/workplace/wuchih/neo-ai-dlr/python/dlr/counter/counter_mgr.py", line 72, in __init__
    raise Exception("unsupported system" + os_name + "," + os_supt)
Exception: unsupported systemLinux,Linux_aarch64
ERROR:root:unsupported system for call home feature
Traceback (most recent call last):
  File "/home/torizon/workplace/wuchih/neo-ai-dlr/python/dlr/counter/counter_mgr.py", line 55, in get_instance
    CallCounterMgr._instance = CallCounterMgr()
  File "/home/torizon/workplace/wuchih/neo-ai-dlr/python/dlr/counter/counter_mgr.py", line 81, in __init__
    raise e
  File "/home/torizon/workplace/wuchih/neo-ai-dlr/python/dlr/counter/counter_mgr.py", line 72, in __init__
    raise Exception("unsupported system" + os_name + "," + os_supt)
Exception: unsupported systemLinux,Linux_aarch64

Reply:
We have just propagate the exception till Call Manger handler function but not intention to stop the application.
The logger by default print exception with trace info. 
We have disabled it now.
About Linux variants system, now we handle that scenario by using default System class Linux_ARM which may also retrieve required information.

Reply:
> About Linux variants system, now we handle that scenario by using default System class Linux_ARM which may also retrieve required information.

I don't think that hard-coded `Linux_ARM` to replace all other machines is a sound approach, though `aarch64` is ARM chip in this case. There are too many variants to validate which machine will work for DLR and which don't.  However, in our case, we would only like to validate that this is linux based system, correct? Since phone-home is an application layer feature, I don't think any machine architecture matter in this case. Hence, I'd still suggest that we allow to send message for Linux base machine. And of course, we can always review the system validation later with @ziyu-guo